### ISSUE TITLE
[subset] Default to all glyphs?

### ISSUE THREAD

**Issue by behdad** [fonttools/fonttools#1909]
A lot of people use our subsetter to simply convert ttf to woff/woff2.  That's a lot of work for such a conversion!  Ideally we'd have a `fonttools.convert` module that wouldn't incur the subsetting cost (and the surprises that come with subsetting) doing that.

But then sometimes people also want to provide `--no-hinting` and other options for such operations.  So yeah, this borders on subsetting, so maybe is fine that subset module is being used for all kinds of conversions.  We can even have a very light `fonttools.convert` module that calls subsetter if necessary...

In such uses it's common for users to pass `--glyphs='*'`.  If no glyph or Unicode codepoints is provided, I'm not sure what the subsetter does currently, but it feels like would be nice defaulting to all glyphs.  Though, one can argue that (and I think I agree!) we should default to all Unicodes and throw out unreachable glyphs.

**Comment by behdad** [comment]
> If no glyph or Unicode codepoints is provided, I'm not sure what the subsetter does currently

We print usage info and bail.

**Comment by BoldMonday** [comment]
We are using `pyftsubset` quite a bit and I appreciate all the efforts that have been put into it.
I guess the reason why people are using it as a woff/woff2 converter has probably to do with zopfli compression. AFAIK no other woff converters offer that.

I am wondering: why throw out glyphs without unicodes by default? For many non-Latin fonts that would break them.

**Comment by m4rc1e** [comment]
@BoldMonday Afaik, pyftsubset by default will [keep](https://github.com/fonttools/fonttools/blob/master/Lib/fontTools/subset/__init__.py#L2310-L2324) non-encoded glyphs which are essential for scripts to work correctly. Instead, it will drop niceties such as stylistic sets etc.

**Comment by behdad** [comment]
> why throw out glyphs without unicodes by default?

That's not what I said. :)  I said "unreachable glyphs".  Those are glyphs that are not accessible via any feature...

Marc is correct that by default the features pyftsubset keeps are features are used by default in shaping.  Ie. any extras are not retained.

**Comment by anthrotype** [comment]
For woff2 there's already `fonttools ttLib.woff2` (but I guess only few people know that).
The `ttx` command can also compile to `--flavor` woff or woff2, and dump woff/woff2 to ttx, so to convert from binary to binary you have to call ttx twice which is not ideal.
A new `fontTools.convert` module and `fonttools convert` command line too could be handy in these circumstances.

As for the subsetter's default options, I often find myself composing very long command lines in order to induce the subsetter to do nothing except what I really want to do (e.g. drop all unreferenced glyphs with --unicodes='*').
I think we have discussed in the past to have a `--profile` option that would switch the set of defaults as a whole. The subset module's option parser is stateful, so one could pass `--profile=keep-everything` (which resets the option defaults to that named profile) and then subsequently on the same command line add other options that modify those default values.

Now we need to at least agree on what such `keep-everything` (or `keep-all`, or a better name) profile should do. Keeping all glyphs is the most conservative choice, even though keeping all unicodes (and dropping unreferenced glypsh) is the most sensible thing to do most of the time.

**Comment by anthrotype** [comment]
that `fonttools convert` script could even be just a short-hand to `fonttools subset --profile=keep-everything`

**Comment by anthrotype** [comment]
changing the current default values for the subsetter is not advisable, in my opinion, as it would break all existing scripts or tools that use it without explicitly setting all values (i.e. a lot)

**Comment by behdad** [comment]
> changing the current default values for the subsetter is not advisable, in my opinion, as it would break all existing scripts or tools that use it without explicitly setting all values (i.e. a lot)

Right.  What I proposed only changes behavior of subset.main() when no glyphs or unicodes are requested.  Currently it just prints usage and ends.  So there won't be any backward compat issue there.

**Comment by anthrotype** [comment]
interesting. Hm I like that actually.

**Comment by anthrotype** [comment]
but then what do we do with the other options' default values (which are very opinionated and geared towards webfont usage). Would the absence of input subset options also change the other default values?
That would be a bit surprising.

**Comment by behdad** [comment]
> Would the absence of input subset options also change the other default values?

No, I'm not suggesting changing default options.  `--profile` sounds good.  Currently internally we differentiate options and "requested glyphs & Unicodes".

Perhaps different profiles will have different defaults for what glyph+unicodes to retain if nothing is specified.  Let's keep thinking, I'm sure something sensible will emerge.
